Add challenge evalenvs

P-Schumacher opened the pull request.

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -37,11 +37,13 @@ class ChallengeOpponent:
     non-disclosed policy will be used.
     """
-    def __init__(self, sim, rng, probabilities: list, min_spawn_distance: float):
+    def __init__(self, sim, rng, probabilities: list, min_spawn_distance: float, linear_velocity: float):
```
Let's make `linear_velocity` consistent with how other variables are passed ie. a tuple `(low, high)`. Depending on how the tuple is passed we can have none `(0,0)` , constant `(vel, vel)`, or random `(0, vel)`

P-Schumacher replied:
imlemented but didnt use `(0,0)` or negative velocities to distinguish between tasks, as we handle that through the `task_type='random'` argument

vikashplus replied:
This might need a close look. 
1. All function signatures must be changed to signify that it's a tuple.
2. It's still being compared as a string throughout. Comparision needs to move to tuple 
3. `task_choice` and `tuple` definition can conflict with each other. Over-specification might be a lot harder to handle.

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -37,11 +37,13 @@ class ChallengeOpponent:
         self.sim = sim
         self.opponent_probabilities = probabilities
         self.min_spawn_distance = min_spawn_distance
+        self.randomize_vel = True if linear_velocity == 'random' else False
```
linear_velocity is passed as float and then treated as string

P-Schumacher replied:
fixed

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -37,11 +37,13 @@ class ChallengeOpponent:
         self.opponent_probabilities = probabilities
         self.min_spawn_distance = min_spawn_distance
+        self.randomize_vel = True if linear_velocity == 'random' else False
+        self.linear_velocity = linear_velocity if type(linear_velocity) == float or type(linear_velocity) == int else 1.0
```
move defaults to the function definitions

P-Schumacher replied:
fixed

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -151,6 +153,10 @@ def reset_opponent(self, player_task='CHASE', rng=None):
 
+        # Randomize opponent forward velocity
+        if self.randomize_vel:
+            self.linear_velocity = self.rng.uniform(1, 5)
```
hard coded - make it an input that can be configured during registration

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -374,6 +380,166 @@ def ncol(self):
+        self.opponent_probabilities = [0.1, 0.35, 0.35, 0.2]
+        self.min_spawn_distance = min_spawn_distance
+        self.noise_process = pink.ColoredNoiseProcess(beta=2, size=(2, 2000), scale=10, rng=rng)
+        self.randomize_vel = True if linear_velocity == 'random' else False
```
follow the same suggestions from above on `linear_velocity`

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -374,6 +380,166 @@ def ncol(self):
+        :param pose: Pose of points along quad boundaries.
+        :type pose: array -> [x, y]
+        """
+        bound_resolution = np.linspace(-8.7, 8.7, 25)
```
hard coded numbers. 
- Best will be to get them automatically 
- At the very least, let's use a global variable for them

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -429,6 +595,8 @@ def _setup(self,
                hills_range=(0,0),
                rough_range=(0,0),
                relief_range=(0,0),
+               velocity=1.0,
```
change it to tuple to specify range 
Can we use sign of the velocity to specify chase / evade?

P-Schumacher replied:
changed to tuple, but didnt use sign to specify chase/evade, see above

vikashplus reviewed (commented):
Left comments inline throughout the PR

P-Schumacher commented:
addressed comments

vikashplus commented on myosuite/envs/myo/myochallenge/__init__.py:
```diff
@@ -76,9 +96,30 @@
             'hills_range': (0.03, 0.23),
             'rough_range': (0.05, 0.1),
             'relief_range': (0.1, 0.3),
+            'velocity': 1.0,
```
Velocity is int here but a str in `myoChallengeChaseTagP2eval-v0`

vikashplus commented on myosuite/envs/myo/myochallenge/__init__.py:
```diff
@@ -76,9 +96,30 @@
+            'hills_range': (0.03, 0.23),
+            'rough_range': (0.05, 0.1),
+            'relief_range': (0.1, 0.3),
+            'velocity': 'random',
```
velocity is float in `myoChallengeChaseTagP1-v0`. Make it consistent

P-Schumacher commented:
Updated settings to have separate velocities for chaser, random and repeller. also added function that asserts correct ranges for these settings.  this should decouple the task choice from the settings.

vikashplus commented on myosuite/envs/myo/myochallenge/__init__.py:
```diff
@@ -41,6 +41,26 @@
+    entry_point='myosuite.envs.myo.myochallenge.relocate_v0:RelocateEnvV0',
+    max_episode_steps=150,
+    kwargs={
+        'model_path': curr_dir+'/../../../simhive/myo_sim/arm/myoarm_object_v0.16(mj237).mjb',
```
using mjb instead of xml

P-Schumacher replied:
switched to xml. that one is still used the procedurally generated objects, is that intentional?

vikashplus commented on myosuite/envs/myo/myochallenge/__init__.py:
```diff
@@ -76,9 +97,35 @@
             'rough_range': (0.05, 0.1),
             'relief_range': (0.1, 0.3),
+            'repeller_opponent': False,
+            'chase_vel_range': (1.0, 1.0), # randomly drawn
```
comments here can be confusing

vikashplus commented on myosuite/envs/myo/myochallenge/__init__.py:
```diff
@@ -76,9 +97,35 @@
+            'repeller_opponent': False,
+            'chase_vel_range': (1.0, 1.0), # randomly drawn
+            'random_vel_range': (-2, 2), # clipped
+            'opponent_probabilities': (0.1, 0.45, 0.45),
```
What is this and why is this of length 4 in `myoChallengeChaseTagP2eval-v0`

P-Schumacher replied:
this specifies the probabilities for the different opponent policies.
0.1 for stationary fixed position
0.45 for stationary random position
0.45 for random moving


There are 4 values for `myoChallengeChaseTagP2eval-v0` because there is also a repeller policy in this case.

I can add an assertion to ensure the right length with `repeller_opponent=True`

P-Schumacher replied:
added assertion

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -37,11 +37,20 @@ class ChallengeOpponent:
+                 probabilities: tuple[float],
+                 min_spawn_distance: float,
+                 chase_vel_range: tuple[float],
+                 random_vel_range: tuple[float],
```
Add comments here to document what these variables are

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -37,11 +37,20 @@ class ChallengeOpponent:
+        self.random_vel_range = random_vel_range
         self.reset_opponent(rng=rng)
 
     def reset_noise_process(self):
```
Do we know if this is deterministic, given seed?

P-Schumacher replied:
yes, the sampling is drawn from the `rng` in the opponent class, which gets set by the outer environment 
- at creation
- every time there is a reset

This ensures that the number generators are always in sync if a seed is set before the episode reset

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -374,6 +386,176 @@ def ncol(self):
+                 random_vel_range: tuple[float],
+                 repeller_vel_range: tuple[float],
+        ):
+        self.dt = 0.01
```
Is there a reason for this to be 10ms. Should it be taken form the env's dt?

P-Schumacher replied:
will change to take it from the env

vikashplus commented on myosuite/envs/myo/myochallenge/chasetag_v0.py:
```diff
@@ -374,6 +386,176 @@ def ncol(self):
+        self.opponent_probabilities = probabilities
+
+        self.min_spawn_distance = min_spawn_distance
+        self.noise_process = pink.ColoredNoiseProcess(beta=2, size=(2, 2000), scale=10, rng=rng)
```
Is this deterministic given seed?

P-Schumacher replied:
yes, the `rng` is passed from the environment and in sync.

vikashplus reviewed (approved):
Looks great. Thanks @P-Schumacher
